R/data-combined.R

```diff
@@ -346,6 +349,144 @@ DataCombined <- R6::R6Class(
       invisible(self)
     },
 
+    #' @description
+    #'
+    #' Get a dataframe for Predicted versus Observed plot
+    #'
+    #' @details
+    #'
+    #' The `threshold` argument can make a big difference to the returned
+    #' dataframe. This argument specifies the tolerance for comparing equality
+    #' of observed and simulated time values (usually found under `xValues`
+    #' column).
+    #'
+    #' Specifically, there are two possibilities to consider. To illustrate
+    #' them, let's say the observed `xValue` is `2` and the simulated values are
+    #' `c(1.4, 2.2, 3.8, 4.0)`.
+    #'
+    #' If the threshold is too **liberal**, multiple simulated values will be
+    #' selected. In such cases, only the first of these is retained. For
+    #' example, with `threshold = 1`, two simulated values (`c(1.4, 2.2`) are
+    #' selected and only the first one (`1.4`) is retained.
+    #'
+    #' On the other hand, if the threshold is too **stringent**, no simulated
+    #' values might be selected. With a `threshold = 0.1`, no simulated values
+    #' will be selected for out example data.
+    #'
+    #' @param threshold Absolute numerical distance by which the closest
+    #'   `xValue` (time, i.e.) in the observed data may differ from `xValue` in
+    #'   simulated data to be accepted as *equal*. If no threshold is specified
+    #'   (default), then the square root of smallest positive floating-point
+    #'   number on your machine will be used (`.Machine$double.eps^0.5`) as a
+    #'   threshold. For all intents and purposes, this amounts to an **exact**
+    #'   match between two values being compared.
+    toObsVsPredDataFrame = function(threshold = NULL) {
+      validateIsNumeric(threshold, nullAllowed = TRUE)
+
+      # We don't want this function to tamper with combined dataframe, so
+      # make an internal copy for this function scope.
+      df <- private$.dataCombined %>%
+        # Remove unnecessary columns
+        private$.cleanDF(.) %>%
+        # Datasets which are not part of any grouping can't be paired
+        dplyr::filter(!is.na(group))
+
+      if (is.null(df)) {
+        stop(
+          "There are currently no datasets. You can add them with `$addDataSets()` and/or `$addSimulationResults()` methods.",
+          call. = FALSE
+        )
+      }
+
+      if (length(unique(df$group)) == 0L) {
+        stop(
+          "There are currently no grouping specified. You can set them with `$setGroups()` method.",
+          call. = FALSE
```

Comment:
why do you always add the call. = FALSE? What does this mean?

Reply:
It decides whether the call should become part of the error.

I always set it to `FALSE` because the error messages are cleaner that way.

``` r
divide1 <- function(x, y) {
  if (y == 0) {
    stop("Can't divide by zero.", call. = TRUE)
  }

  x / y
}

divide1(1, 0)
#> Error in divide1(1, 0): Can't divide by zero.

divide2 <- function(x, y) {
  if (y == 0) {
    stop("Can't divide by zero.", call. = FALSE)
  }

  x / y
}

divide2(1, 0)
#> Error: Can't divide by zero.
```

<sup>Created on 2022-03-07 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

We can create a helper in _ospsuite.utils_ to avoid this duplication:

```r
stopWithoutCall <- function(...) stop(..., call. = FALSE)
```

What do you think?

Reply:
I prefer to have the call in the message, as it makes debugging easier. It does not matter in your example as the error is thrown by the direct call, but in case of an error being thrown from a function or an object's method, `call. = FALSE` is just not informative enough.

Reply:
Because, typically, the errors bubble up from somewhere deep in the code, and we can't expect the user to be familiar with our _implementation_. Our error messages tend to be verbose and insightful enough for them to know what went wrong. And, if the user is savvy enough, they can just call `traceback()` and it will show them where the error is originating from.

It might seem helpful in this toy example to show the call, but imagine the more realistic scenario. 

Let's say I run a function called `computeEllipsisCircumference()` and it errors with `#> Error in divide1(1, 0): Can't divide by zero`. 

As a user, how is `divide1(1, 0)` call useful?

Reply:
> As a user, how is divide1(1, 0) call useful?

To be honest, at this state I am more concernd about me as the dev, and `divide1(1, 0)` would definitely help me to find out where I have to debug.

Reply:
As a developer, can't you just run `traceback()`?

Reply:
I would like to have the function call in the error message.

Reply:
Since Juri and you both seem to find the function call in error message useful, I have reverted this change.

The error messages will now show the function call.

# before

``` r
library(ospsuite)

myCombDat <- DataCombined$new()
myCombDat$toObsVsPredDataFrame()
#> Error: There are currently no datasets. You can add them with `$addDataSets()` and/or `$addSimulationResults()` methods.
```

<sup>Created on 2022-03-07 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

# after

``` r
library(ospsuite)
#> Loading required package: rClr
#> Loading the dynamic library for Microsoft .NET runtime...
#> Loaded Common Language Runtime version 4.0.30319.42000

myCombDat <- DataCombined$new()
myCombDat$toObsVsPredDataFrame()
#> Error in myCombDat$toObsVsPredDataFrame(): There are currently no datasets. You can add them with `$addDataSets()` and/or `$addSimulationResults()` methods.
```

<sup>Created on 2022-03-07 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

Reply:
https://adv-r.hadley.nz/conditions.html#signalling-conditions

<img width="1211" alt="Screenshot 2022-03-24 at 23 33 22" src="https://user-images.githubusercontent.com/11330453/160021176-95308ad4-4f7e-46be-a504-b500df2d553e.png">

Reply:
I am just thinking about the situation where user reports a bug.
With `call=FALSE` we will not immediately get the call stack and will need to ask the user to rerun the workflow, call after the  `traceback()` and send us the error call stack.

Reply:
I think we already agreed on having `.call = TRUE`.